js/compiler.ts

```diff
@@ -0,0 +1,502 @@
+    target: ts.ScriptTarget.ESNext
+  };
+
+  private _os: Os = os;
+  private _service: ts.LanguageService;
+  private _ts: Ts = ts;
+
+  /**
+   * Setup the global error handler for privileged bindings.
+   */
+  private _setupGlobalErrorHandler(): void {
+    // Uncaught exceptions are sent to window.onerror by the privileged binding.
+    window.onerror = (
+      message: string,
+      source: string,
+      lineno: number,
+      colno: number,
+      error: Error
+    ) => {
+      // TODO Currently there is a bug in v8_source_maps.ts that causes a
+      // segfault if it is used within window.onerror. To workaround we
+      // uninstall the Error.prepareStackTrace handler. Users will get unmapped
+      // stack traces on uncaught exceptions until this issue is fixed.
+      //Error.prepareStackTrace = null;
+      console.log(error.stack);
+      this._os.exit(1);
+    };
+  }
+
+  /**
+   * Setup being able to map back source references back to their source
+   */
+  private _setupSourceMaps(): void {
+    sourceMaps.install({
+      installPrepareStackTrace: true,
+      getGeneratedContents: (fileName: string): string | RawSourceMap => {
+        log("getGeneratedContents", fileName);
+        if (fileName === "gen/bundle/main.js") {
+          assert(libdeno.mainSource.length > 0);
+          return libdeno.mainSource;
+        } else if (fileName === "main.js.map") {
+          return libdeno.mainSourceMap;
+        } else if (fileName === "deno_main.js") {
+          return "";
+        } else {
+          const moduleFile = this._fileModuleMap.get(fileName);
+          if (!moduleFile) {
+            log("getGeneratedContents cannot find", fileName);
+            return "";
+          }
+          return moduleFile.outputCode;
+        }
+      }
+    });
+  }
+
+  private constructor() {
+    if (DenoCompiler._instance) {
+      throw new TypeError("Attempt to create an additional compiler.");
+    }
+    this._service = this._ts.createLanguageService(this);
+    this._setupGlobalErrorHandler();
+    this._setupSourceMaps();
+  }
+
+  // Deno specific compiler API
+
+  /**
+   * Retrieve the output of the TypeScript compiler for a given `fileName`.
+   */
+  compile(fileName: string): string {
+    const service = this._service;
+    const output = service.getEmitOutput(fileName);
+
+    // Get the relevant diagnostics - this is 3x faster than
+    // `getPreEmitDiagnostics`.
+    const diagnostics = [
+      ...service.getCompilerOptionsDiagnostics(),
+      ...service.getSyntacticDiagnostics(fileName),
+      ...service.getSemanticDiagnostics(fileName)
+    ];
+    if (diagnostics.length > 0) {
+      const errMsg = this._ts.formatDiagnosticsWithColorAndContext(
+        diagnostics,
+        formatDiagnosticsHost
+      );
+      console.log(errMsg);
+      this._os.exit(1);
+    }
+
+    assert(!output.emitSkipped, "The emit was skipped for an unknown reason.");
+
+    return output.outputFiles[0].text;
```

Comment:
I believe `output.outputFiles[1]` would contain the external source map if there was one - so I think you can add a `assert(output.outputFiles.length == 1)`. If that's possible please do that - and add a comment to the effect of:
```typescript
// There's only one outputFile because inlineSourceMap is false.
// See issue https://github.com/denoland/deno/issues/23
assert(output.outputFiles.length == 1);
```